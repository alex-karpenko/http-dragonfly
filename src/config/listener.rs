use serde::{
    de::{self, Visitor},
    Deserialize, Deserializer,
};
use std::{fmt::Display, net::Ipv4Addr, str::FromStr, time::Duration};

use super::{headers::HeaderTransform, response::ResponseConfig, target::TargetConfig};

const DEFAULT_LISTENER_PORT: u16 = 8080;
pub const DEFAULT_LISTENER_TIMEOUT_SEC: u64 = 10;
const INVALID_IP_ADDRESS_ERROR: &str = "IP address isn't valid";

#[derive(Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct ListenerConfig {
    name: Option<String>,
    #[serde(rename = "on", default)]
    listen_on: ListenOn,
    #[serde(
        with = "humantime_serde",
        default = "ListenerConfig::default_listener_timeout"
    )]
    timeout: Duration,
    headers: Option<Vec<HeaderTransform>>,
    methods: Option<Vec<String>>,
    targets: Vec<TargetConfig>,
    #[serde(default)]
    response: ResponseConfig,
}

impl ListenerConfig {
    fn default_listener_timeout() -> Duration {
        Duration::from_secs(DEFAULT_LISTENER_TIMEOUT_SEC)
    }

    pub fn get_name(self) -> String {
        if let Some(name) = self.name {
            name
        } else {
            format!("LISTENER-{}", self.listen_on)
        }
    }
}

#[derive(Deserialize, Debug)]
#[serde(deny_unknown_fields, rename_all = "lowercase")]
pub enum HttpMethod {
    Get,
    Post,
    Put,
    Patch,
    Delete,
    Options,
    Head,
}

#[derive(Debug)]
pub struct ListenOn {
    ip: Ipv4Addr,
    port: u16,
}

impl Default for ListenOn {
    fn default() -> Self {
        Self {
            ip: Ipv4Addr::new(0, 0, 0, 0),
            port: DEFAULT_LISTENER_PORT,
        }
    }
}

impl ListenOn {
    fn parse_ip_address(ip: &str) -> Result<Ipv4Addr, String> {
        Ipv4Addr::from_str(ip).map_err(|_| String::from(INVALID_IP_ADDRESS_ERROR))
    }

    fn from_str(v: &str) -> Result<Self, String> {
        let splitted: Vec<_> = v.trim().split(':').collect();

        if splitted.len() == 1 {
            let port: u16 = splitted[0]
                .parse()
                .map_err(|e| format!("invalid port value `{}`: {e}", splitted[0]))?;
            let ip = Ipv4Addr::new(0, 0, 0, 0);

            Ok(ListenOn { ip, port })
        } else if splitted.len() == 2 {
            let port: u16 = splitted[1]
                .parse()
                .map_err(|e| format!("invalid port value `{}`: {e}", splitted[1]))?;

            let ip = if splitted[0].is_empty() || splitted[0] == "*" {
                Ipv4Addr::new(0, 0, 0, 0)
            } else {
                Self::parse_ip_address(splitted[0])?
            };

            Ok(ListenOn { ip, port })
        } else {
            Err("invalid `listen on` token, should be in form IP:PORT".into())
        }
    }
}

impl Display for ListenOn {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:{}", self.ip, self.port)
    }
}

impl<'de> Deserialize<'de> for ListenOn {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct ListenOnVisitor;
        impl<'de> Visitor<'de> for ListenOnVisitor {
            type Value = ListenOn;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                formatter.write_str("an IP address (or `0.0.0.0` or `*`) and port separated by colon, like `1.2.3.4:8080`")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                ListenOn::from_str(v).map_err(|e| E::custom(e))
            }
        }

        deserializer.deserialize_string(ListenOnVisitor)
    }
}
