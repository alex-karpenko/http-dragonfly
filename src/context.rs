use hyper::{http::request::Parts, Body, Response};
use once_cell::sync::OnceCell;
use regex::Regex;
use std::collections::HashMap;
use std::env;
use std::net::SocketAddr;

use crate::config::target::TargetConfig;

const CTX_APP_NAME: &str = env!("CARGO_PKG_NAME");
const CTX_APP_VERSION: &str = env!("CARGO_PKG_VERSION");

pub type ContextMap = HashMap<String, String>;

#[derive(Debug)]
pub struct Context<'a> {
    own: ContextMap,
    parent: Option<&'a Context<'a>>,
}

impl<'a> Context<'a> {
    pub fn root(root_env: impl RootEnvironment) -> &'static Context<'a> {
        static ROOT_CONTEXT: OnceCell<Context> = OnceCell::new();

        let ctx = ROOT_CONTEXT.get_or_init(|| {
            let mut ctx = ContextMap::new();

            ctx.insert("CTX_APP_NAME".into(), CTX_APP_NAME.into());
            ctx.insert("CTX_APP_VERSION".into(), CTX_APP_VERSION.into());
            ctx.extend(root_env.get_environment());

            Context {
                own: ctx,
                parent: None,
            }
        });

        ctx
    }

    pub fn with(&self, own: ContextMap) -> Context {
        Context {
            own,
            parent: Some(self),
        }
    }

    pub fn get(&self, var: &String) -> Option<&String> {
        // Try own context
        if let Some(value) = self.own.get(var) {
            Some(value)
            // or get parent if it's
        } else if let Some(parent) = &self.parent {
            // and try parent
            if let Some(value) = parent.get(var) {
                Some(value)
            } else {
                None
            }
        } else {
            None
        }
    }

    pub fn iter(&'a self) -> Box<dyn Iterator<Item = (&'a String, &'a String)> + 'a> {
        let iter = Box::new(self.own.iter());
        Box::new(Iter {
            ctx: self,
            iter,
            finished: false,
        })
    }

    pub fn with_request(
        &'a self,
        addr: &'a SocketAddr,
        req: &'a Parts,
        listener_name: String,
    ) -> Context<'a> {
        let mut own = ContextMap::new();

        // CTX_LISTENER_NAME
        // CTX_REQUEST_SOURCE_IP
        // CTX_REQUEST_METHOD
        // CTX_REQUEST_HOST
        // CTX_REQUEST_PATH
        // CTX_REQUEST_QUERY
        own.insert("CTX_LISTENER_NAME".into(), listener_name);
        own.insert("CTX_REQUEST_SOURCE_IP".into(), addr.ip().to_string());
        own.insert("CTX_REQUEST_METHOD".into(), req.method.to_string());
        own.insert("CTX_REQUEST_PATH".into(), req.uri.path().to_string());
        if let Some(host) = req.uri.host() {
            own.insert("CTX_REQUEST_HOST".into(), host.to_lowercase());
        }
        if let Some(query) = req.uri.query() {
            own.insert("CTX_REQUEST_QUERY".into(), query.to_lowercase());
        }

        // CTX_REQUEST_HEADERS_<UPPERCASE_HEADER_NAME>
        req.headers.iter().for_each(|(n, v)| {
            let n = n.as_str().to_uppercase().replace('-', "_");
            let v = v.to_str().unwrap_or("").to_string();
            own.insert(format!("CTX_REQUEST_HEADERS_{n}"), v);
        });

        self.with(own)
    }

    pub fn with_target(&'a self, cfg: &'a TargetConfig) -> Context<'a> {
        let mut own = ContextMap::new();

        // CTX_TARGET_ID
        // CTX_TARGET_HOST
        own.insert("CTX_TARGET_ID".into(), cfg.id());
        own.insert("CTX_TARGET_HOST".into(), cfg.host());

        self.with(own)
    }

    pub fn with_response(&'a self, resp: &Response<Body>) -> Context<'a> {
        let mut own = ContextMap::new();

        // CTX_RESPONSE_HEADERS_<UPPERCASE_HEADER_NAME>
        // CTX_RESPONSE_STATUS
        own.insert("CTX_RESPONSE_STATUS".into(), resp.status().to_string());
        resp.headers().iter().for_each(|(n, v)| {
            let n = n.as_str().to_uppercase().replace('-', "_");
            let v = v.to_str().unwrap_or("").to_string();
            own.insert(format!("CTX_RESPONSE_HEADERS_{n}"), v);
        });

        self.with(own)
    }
}

pub struct Iter<'a> {
    ctx: &'a Context<'a>,
    iter: Box<dyn Iterator<Item = (&'a String, &'a String)> + 'a>,
    finished: bool,
}

impl<'a> Iterator for Iter<'a> {
    type Item = (&'a String, &'a String);

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(next) = self.iter.next() {
            Some(next)
        } else if self.ctx.parent.is_none() {
            None
        } else if !self.finished {
            self.finished = true;
            self.iter = self.ctx.parent.unwrap().iter();
            self.iter.next()
        } else {
            None
        }
    }
}

pub trait RootEnvironment {
    fn get_environment(&self) -> ContextMap;
}

pub struct RootOsEnvironment<'a> {
    env_mask_regex: &'a str,
}

impl<'a> RootEnvironment for RootOsEnvironment<'a> {
    fn get_environment(&self) -> ContextMap {
        let mut env_ctx = ContextMap::new();
        let re = Regex::new(self.env_mask_regex).expect("invalid environment filter regex");
        env::vars().for_each(|(k, v)| {
            if re.is_match(&k) {
                env_ctx.insert(k, v);
            }
        });

        env_ctx
    }
}

impl<'a> RootOsEnvironment<'a> {
    pub fn new(env_mask_regex: &'a str) -> Self {
        Self { env_mask_regex }
    }
}
