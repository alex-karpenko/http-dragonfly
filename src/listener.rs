use futures_util::future::join_all;
use http::{request::Parts, Error, HeaderValue};
use hyper::{
    header::HOST, http, Body, Client, Error as HyperError, HeaderMap, Request, Response,
    StatusCode, Uri,
};
use shellexpand::env_with_context_no_errors;
use std::{net::SocketAddr, collections::HashMap};
use tracing::debug;

use crate::{
    config::{
        headers::{HeaderTransform, HeaderTransformActon},
        listener::ListenerConfig,
        response::{OverrideConfig, ResponseStatus, ResponseStrategy},
        target::{TargetConfig, TargetOnErrorAction},
    },
    context::{Context, ContextMap},
};

pub struct Listener {}

impl Listener {
    pub async fn handler<'a>(
        cfg: &'static ListenerConfig,
        ctx: &'static Context<'a>,
        addr: SocketAddr,
        req: Request<Body>,
    ) -> Result<Response<Body>, Error> {
        // Verify is method allowed in the config
        if !cfg.is_method_allowed(req.method().as_ref()) {
            debug!("method `{}` rejected", req.method().to_string());
            return Response::builder()
                .status(StatusCode::METHOD_NOT_ALLOWED)
                .body(Body::empty());
        }

        // Prepare owned clonable body
        let (req_parts, req_body) = req.into_parts();
        let body_bytes = hyper::body::to_bytes(req_body).await.unwrap();
        // Add own context - listener + request
        let ctx = Listener::request_context(cfg, ctx, &addr, &req_parts);
        //debug!("request context: {:?}", ctx);

        // Prepare new headers
        let mut headers = req_parts.headers.clone();
        headers.remove(HOST);
        if let Some(transforms) = &cfg.headers {
            Listener::transform_headers(&mut headers, transforms, &ctx);
        }
        debug!("request headers: {:?}", headers);

        // Process targets
        debug!(
            "Listener={}, strategy={}",
            cfg.get_name(),
            cfg.response.strategy
        );

        let mut target_requests = vec![];
        let mut target_ctx = vec![];
        let mut target_ids: HashMap<String, usize> = HashMap::new();

        let http_client = Client::new();

        for (pos, target) in cfg.targets.iter().enumerate() {
            let ctx = Listener::target_context(&target, &ctx);
            let target_request_builder = Request::builder();
            // Set method
            let target_request_builder = target_request_builder.method(&req_parts.method);
            // Set uri
            let url = env_with_context_no_errors(&target.url, |v| ctx.get(&v.into()));
            let uri: Uri = url.parse()?;
            let mut target_request_builder = target_request_builder.uri(uri);
            // Prepare headers
            let mut headers = headers.clone();
            if let Some(transforms) = &target.headers {
                Listener::transform_headers(&mut headers, transforms, &ctx);
            }
            // Add Host header if empty
            if !headers.contains_key(HOST) {
                debug!("add host header");
                let host: Uri = url.parse()?;
                let host = host.host().unwrap();
                headers.insert(HOST, HeaderValue::from_str(host)?);
            }
            // Insert all headers into request
            for (k, v) in &headers {
                target_request_builder = target_request_builder.header(k, v);
            }
            // Finalize request with body
            let target_request: Request<Body> = if let Some(body) = &target.body {
                let body = env_with_context_no_errors(body, |v| ctx.get(&v.into()));
                target_request_builder.body(Body::from(body))?
            } else {
                target_request_builder.body(Body::from(body_bytes.clone()))?
            };

            // Put request to queue
            debug!("target `{}` request: {:?}", target.get_id(), target_request);
            target_requests.push(http_client.request(target_request));
            target_ctx.push(ctx);
            target_ids.insert(target.get_id(), pos);
        }

        // Get results
        let results: Vec<Result<Response<Body>, HyperError>> = join_all(target_requests).await;
        // Pre-process results
        let mut responses = vec![];
        for (pos, res) in results.into_iter().enumerate() {
            match res {
                Ok(resp) => {
                    debug!("OK: {:#?}", resp);
                    responses.push(Some(resp))
                }
                Err(e) => {
                    debug!("ERR: {:#?}", e);
                    let target = &cfg.targets[pos];
                    let resp = match target.on_error {
                        TargetOnErrorAction::Propagate => {
                            Some(Listener::build_on_error_response(e, &target.error_status))
                        }
                        TargetOnErrorAction::Status => {
                            Some(Listener::build_on_error_response(e, &target.error_status))
                        }
                        TargetOnErrorAction::Drop => None,
                    };
                    responses.push(resp);
                }
            }
        }

        // Select response according to strategy
        let resp = match &cfg.response.strategy {
            ResponseStrategy::AlwaysOverride => Listener::override_response(
                Response::new(Body::empty()),
                &ctx,
                &cfg.response.override_config,
            ),
            ResponseStrategy::AlwaysTargetId => todo!(),
            ResponseStrategy::OkThenFailed => todo!(),
            ResponseStrategy::OkThenTargetId => todo!(),
            ResponseStrategy::OkThenOverride => todo!(),
            ResponseStrategy::FailedThenOk => todo!(),
            ResponseStrategy::FailedThenTargetId => todo!(),
            ResponseStrategy::FailedThenOverride => todo!(),
            ResponseStrategy::ConditionalRouting => todo!(),
        };

        // Final response
        Ok(resp)
    }

    fn request_context<'a>(
        cfg: &'a ListenerConfig,
        ctx: &'a Context,
        addr: &'a SocketAddr,
        req: &'a Parts,
    ) -> Context<'a> {
        let mut own = ContextMap::new();

        // CTX_LISTENER_NAME
        // CTX_REQUEST_SOURCE_IP
        // CTX_REQUEST_METHOD
        // CTX_REQUEST_HOST
        // CTX_REQUEST_PATH
        // CTX_REQUEST_QUERY
        own.insert("CTX_LISTENER_NAME".into(), cfg.get_name());
        own.insert("CTX_REQUEST_SOURCE_IP".into(), addr.ip().to_string());
        own.insert("CTX_REQUEST_METHOD".into(), req.method.to_string());
        own.insert("CTX_REQUEST_PATH".into(), req.uri.path().to_string());
        if let Some(host) = req.uri.host() {
            own.insert("CTX_REQUEST_HOST".into(), host.to_lowercase().to_string());
        }
        if let Some(query) = req.uri.query() {
            own.insert("CTX_REQUEST_QUERY".into(), query.to_lowercase().to_string());
        }

        // CTX_REQUEST_HEADERS_<UPPERCASE_HEADER_NAME>
        req.headers.iter().for_each(|(n, v)| {
            let n = n.as_str().to_uppercase().replace('-', "_");
            let v = v.to_str().unwrap_or("").to_string();
            own.insert(format!("CTX_REQUEST_HEADERS_{n}"), v);
        });

        ctx.with(own)
    }

    fn target_context<'a>(cfg: &'a TargetConfig, ctx: &'a Context) -> Context<'a> {
        let mut own = ContextMap::new();

        // CTX_TARGET_ID
        // CTX_TARGET_HOST
        own.insert("CTX_TARGET_ID".into(), cfg.get_id());
        own.insert(
            "CTX_TARGET_HOST".into(),
            cfg.get_uri()
                .unwrap()
                .host()
                .unwrap()
                .to_lowercase()
                .to_string(),
        );

        ctx.with(own)
    }

    fn transform_headers(
        headers: &mut HeaderMap,
        transforms: &'static Vec<HeaderTransform>,
        ctx: &Context,
    ) {
        for transform in transforms {
            match &transform.action {
                HeaderTransformActon::Add(key) => {
                    if !headers.contains_key(key) {
                        let value = transform.value.as_ref().unwrap().as_str();
                        let value = env_with_context_no_errors(value, |v| ctx.get(&v.into()));
                        headers.insert(key.as_str(), HeaderValue::from_str(&value).unwrap());
                    }
                }
                HeaderTransformActon::Replace(key) => {
                    let value = transform.value.as_ref().unwrap().as_str();
                    let value = env_with_context_no_errors(value, |v| ctx.get(&v.into()));
                    headers.insert(key.as_str(), HeaderValue::from_str(&value).unwrap());
                }
                HeaderTransformActon::Drop(key) => {
                    if key == "*" {
                        headers.clear();
                    } else {
                        headers.remove(key);
                    }
                }
            };
        }
    }

    fn build_on_error_response(e: HyperError, status: &Option<ResponseStatus>) -> Response<Body> {
        let resp = Response::builder();

        let resp = if let Some(status) = status {
            resp.status(status.get_code())
        } else if e.is_connect() || e.is_closed() {
            resp.status(502)
        } else if e.is_timeout() {
            resp.status(504)
        } else {
            resp.status(500)
        };

        resp.body(Body::empty()).unwrap()
    }

    fn override_response(
        resp: Response<Body>,
        ctx: &Context,
        cfg: &'static Option<OverrideConfig>,
    ) -> Response<Body> {
        if let Some(cfg) = cfg {
            let (resp_parts, resp_body) = resp.into_parts();
            let mut new_resp = Response::builder();

            // Set status
            new_resp = if let Some(status) = &cfg.status {
                new_resp.status(status.get_code())
            } else {
                new_resp.status(resp_parts.status)
            };

            // Prepare headers
            let mut headers = resp_parts.headers.clone();
            //let cfg_headers = cfg.headers.clone();
            if let Some(transforms) = &cfg.headers {
                Listener::transform_headers(&mut headers, transforms, ctx);
            }
            for (k, v) in &headers {
                new_resp = new_resp.header(k, v);
            }

            // Prepare body
            let cfg_body = cfg.body.clone();
            let body: Body = if let Some(body) = cfg_body {
                let body: String = env_with_context_no_errors(&body, |v| ctx.get(&v.into())).into();
                Body::from(body)
            } else {
                resp_body
            };

            // Final response
            new_resp.body(body).unwrap()
        } else {
            resp
        }
    }
}
