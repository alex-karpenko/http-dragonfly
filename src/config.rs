pub mod headers;
pub mod listener;
pub mod response;
pub mod strategy;
pub mod target;

use figment::{
    providers::{Format, Yaml},
    Figment,
};
use once_cell::sync::OnceCell;
use serde::Deserialize;
use shellexpand::env_with_context_no_errors;
use std::fs::read_to_string;
use tracing::{debug, info};

use crate::{context::Context, errors::HttpDragonflyError};

use self::listener::ListenerConfig;

static APP_CONFIG: OnceCell<AppConfig> = OnceCell::new();

pub trait ConfigValidator {
    fn validate(&self) -> Result<(), HttpDragonflyError>;
}

#[derive(Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct AppConfig {
    listeners: Vec<ListenerConfig>,
}

impl<'a> AppConfig {
    pub fn new(filename: &String, ctx: &Context) -> Result<&'a AppConfig, HttpDragonflyError> {
        let config = AppConfig::owned(filename, ctx)?;
        Ok(APP_CONFIG.get_or_init(|| config))
    }

    fn owned(filename: &String, ctx: &Context) -> Result<AppConfig, HttpDragonflyError> {
        info!("Loading config: {filename}");
        let config = read_to_string(filename).map_err(|e| HttpDragonflyError::LoadConfigFile {
            filename: filename.clone(),
            cause: e,
        })?;
        let config = env_with_context_no_errors(&config, |v| ctx.get(&v.into()));
        let config: AppConfig = Figment::new().merge(Yaml::string(&config)).extract()?;

        debug!("Application config: {:#?}", config);
        match config.validate() {
            Ok(_) => Ok(config),
            Err(e) => Err(e),
        }
    }

    pub fn listeners(&self) -> &[ListenerConfig] {
        self.listeners.as_ref()
    }
}

impl ConfigValidator for AppConfig {
    fn validate(&self) -> Result<(), HttpDragonflyError> {
        for listener in self.listeners() {
            listener.validate()?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use crate::context::test_context;
    use insta::{assert_debug_snapshot, glob};

    use super::*;

    const TEST_CONFIGS_FOLDER: &str = "../tests/configs";

    #[test]
    fn good_config() {
        let ctx = test_context::get_test_ctx();
        glob!(
            TEST_CONFIGS_FOLDER,
            "good/*.yaml",
            |path| assert_debug_snapshot!(AppConfig::owned(
                &String::from(path.to_str().unwrap()),
                &ctx
            ))
        );
    }

    #[test]
    fn wrong_config() {
        let ctx = test_context::get_test_ctx();
        glob!(
            TEST_CONFIGS_FOLDER,
            "wrong/*.yaml",
            |path| insta::with_settings!({filters => vec![(r#"unable to parse config: invalid config: found "(.+)" but expected one of (.+),"#, "unable to parse config: invalid config: found [SOMETHING] but expected one of [EXPECTED JQ STATEMENTS]")]},
                {assert_debug_snapshot!(AppConfig::owned(&String::from(path.to_str().unwrap()),&ctx));})
        );
    }
}
