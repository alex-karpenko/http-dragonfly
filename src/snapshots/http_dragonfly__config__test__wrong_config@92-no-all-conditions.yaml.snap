---
source: src/config.rs
expression: "AppConfig::owned(&String::from(path.to_str().unwrap()), &ctx)"
input_file: tests/configs/wrong/92-no-all-conditions.yaml
---
Err(
    invalid config: all targets must have condition defined because strategy is `conditional_routing`,
)
